torch/distributed/pipelining/schedules.py

```diff
@@ -2097,71 +2166,331 @@ def backward_stage_index(step):
-            return False
 
-        seen_ops: Set[Tuple[int, _ComputationType, int]] = set()
-        result: Dict[int, List[Optional[_Action]]] = {}
-        next_pointer: Dict[int, int] = {}
-        bubbles_added: Dict[int, int] = {}
-        total_bubbles_added = 0
 
-        for rank in range(self.pp_group_size):
-            result[rank] = []
-            next_pointer[rank] = 0
-            bubbles_added[rank] = 0
+class ScheduleZBVZeroBubble(PipelineScheduleMulti):
+    """
+    The ZBV Zero Bubble schedule.
+    See https://arxiv.org/pdf/2401.10241 for details.
+    """
 
-        while True:
-            should_stop = True
+    def __init__(
+        self,
+        stages: List[_PipelineStageBase],
+        n_microbatches: int,
+        loss_fn: Optional[Callable] = None,
+        args_chunk_spec: Optional[Tuple[TensorChunkSpec, ...]] = None,
+        kwargs_chunk_spec: Optional[Dict[str, TensorChunkSpec]] = None,
+        output_merge_spec: Optional[Union[Dict[str, Any], Tuple[Any]]] = None,
+    ):
+        self.pp_group_size = stages[0].group_size
+        super().__init__(
+            stages=stages,
+            n_microbatches=n_microbatches,
+            loss_fn=loss_fn,
+            args_chunk_spec=args_chunk_spec,
+            kwargs_chunk_spec=kwargs_chunk_spec,
+            output_merge_spec=output_merge_spec,
+            use_full_backward=False,
+        )
+        self.n_local_stages = len(stages)
+        self.rank = stages[0].group_rank
+        if self.n_local_stages % 2 != 0:
+            raise ValueError(
+                "ZBV schedule requires the number of local chunks to be "
+                f"divisible by 2, but got {len(stages)}."
+            )
 
-            temp_seen_ops: Set[Tuple[int, _ComputationType, int]] = set()
+        # 1. Create the pipeline_order (all ranks do this calculation)
+        # This will be used to keep track of the current state of the entire pipeline
+        # pipeline_order[rank] = [Action(computation_type, microbatch_index, stage_index), ...]
+        self.pipeline_order: Dict[int, List[Optional[_Action]]] = self._get_zbv_schedule(
+            len(stages),
+            self.pp_group_size,
+            n_microbatches,
+        )
+        self.pipeline_order = _add_bubbles_to_actions(
+            self.n_local_stages * self.pp_group_size,
+            self.pipeline_order,
+            self.pp_group_size,
+        )
 
-            for rank in range(self.pp_group_size):
-                timestamp = next_pointer[rank]
-                if timestamp >= len(actions[rank]):
+    def _get_zbv_schedule(
+        self,
+        num_model_chunks,
+        pipeline_parallel_size,
+        num_microbatches,
+    ):
+        pp_group_size = pipeline_parallel_size
+        mb_size = num_microbatches
+        num_local_stages = num_model_chunks
+        compute_schedules = {}
+        category_map = {
+            F: 0,
+            I: 1,
+            W: 2,
+        }
+        chunk_0 = 0
```

Comment:
does zbv hardcode that there are exactly 2 model-chunks per PP rank? should we make that more of an assertion if so?